assembler/src/lib.rs

```diff
@@ -0,0 +1,74 @@
+mod backend;
+
+use libc;
+use std::mem;
+use std::process::Command;
+
+// getpagesize() (sycall)
+const PAGE_SIZE: usize = 4096;
+
+struct Assembler {
+    memmap: *mut u8,
+    instrptr: usize
+}
+
+impl Assembler {
+
+    pub fn new() -> Self {
+        // Create a single page of memory to store our instructions in and make it writable
+        let memmap = unsafe {
+            let mut page: *mut libc::c_void = mem::MaybeUninit::uninit().as_mut_ptr();
+            libc::posix_memalign(&mut page, PAGE_SIZE, PAGE_SIZE);
+            libc::mprotect(page, PAGE_SIZE, libc::PROT_WRITE);
+            // Transmute to raw u8 pointer
+            let memmap: *mut u8 = mem::transmute(page);
+            // TODO correct use of MaybeUninit::assume_init()
+            memmap
+        };
+
+        Assembler {
+            memmap,
+            instrptr: 0
+        }
+    }
+
+    pub fn add_instruction(&mut self, instr: &str) {
+        let mc = backend::encode(backend::EncoderBackend::RasmBackend, instr);
+        unsafe {
+            let off = self.memmap.offset(self.instrptr as isize);
+            libc::memcpy(off as *mut _, mc.as_ptr() as *const _, mc.len());
+        }
+        self.instrptr += mc.len();
+    }
+
+    fn make_executable(&self) {
+        unsafe {
+            libc::mprotect(self.memmap as *mut _, PAGE_SIZE, libc::PROT_EXEC);
+        }
+    }
+
+    fn function_pointer(&self) -> fn() -> i64 {
+        unsafe {
+            mem::transmute(self.memmap)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::Assembler;
+
+    #[test]
+    pub(crate) fn test_basic() {
+        let mut assemb = Assembler::new();
```

Comment:
I think we can separate out the "building instructions" phase from "something that's executable" phase? So we might end up with something like:

```
let mut asm = AsmBuilder::new();
asm.push("mov rdx, 3");
asm.push("mov rcx, 4");
...
let f = asm.assemble();
assert_eq!(f(), 7);
```

Where the type structure might look something like:

```
struct AsmBuilder { ... }

impl AsmBuilder {
    fn push(&mut self, instr: &str) { ... }
    fn assemble(self) -> Assembled { ... }
}

struct AssembledFunction { ... }

impl AssembledFunction {
    fn as_ptr(&self) -> ...
}
```

Note that `assemble` takes `self` (not `&self`): when you call "assemble" you would consume the `AsmBuilder`. There are some advantages to this:

* We can choose whether we convert strings to instructions on each `push` or at `assemble` (the latter stage allows us to do things with labels etc.).
* Once `assemble` is called, people can't add instructions to a page of memory marked as executable. This will make us more easily compatible with W^X.

I can't think of any immediate disadvantages. We could even have a way of consuming an `AssembledFunction` back to an `AsmBuilder` for when we want to patch code. But that's a little way off!

Reply:
I'm in favour of assembling as we go rather than caching strings. The binary encodings are going to be smaller, thus less memory to store.

Consuming self sounds good to me.

Reply:
> I'm in favour of assembling as we go rather than caching strings.

The API I've suggested doesn't expose to users which we do.

Reply:
I like this. I assume the `AssembledFunction` would hold on to the memory map as well, so that it doesn't go out of scope. Is there anything else that `AssemblerFunction` could do apart from giving us a pointer?

Reply:
> The API I've suggested doesn't expose to users which we do.

Yes

Reply:
> I assume the AssembledFunction would hold on to the memory map as well, so that it doesn't go out of scope

Yes, we can add a `Drop` to `AssembledFunction` that unmaps and frees the memory automatically.

> Is there anything else that AssemblerFunction could do apart from giving us a pointer?

I don't know what else it might need to do, but I haven't thought deeply about use cases here.

Reply:
> Is there anything else that AssemblerFunction could do apart from giving us a pointer?

Eventually it'll have to return a signature of the code it generated, but I think we can assume a fixed sig for now. We can build up to that later.

Reply:
> Eventually it'll have to return a signature of the code it generated

I can't think of anyway of encoding this statically. I think ultimately we give users some sort of pointer back and they have to cast it (with `unsafe`) to the correct Rust/FFI function type.